src/lib/mixer/mixer_multirotor.cpp

```diff
@@ -203,48 +200,38 @@ MultirotorMixer::mix(float *outputs, unsigned space)
 
 	float boost = 0.0f;		// value added to demanded thrust (can also be negative)
 	float roll_pitch_scale = 1.0f;	// scale for demanded roll and pitch
-
-	// Note: thrust boost is computed assuming thrust_gain==1 for all motors.
-	// On asymmetric platforms, some motors have thrust_gain<1,
+	float delta_out_max = max_out - min_out; // distance between the two extrema
+
+	// If the difference between the to extrema is smaller than 1.0, the boost can safely unsaturate a motor if needed
+	// without saturating another one.
+	// Otherwise, a scaler is computed to make the distance between the two extrema exacly 1.0 and the boost
+	// value is computed to maximize the roll-pitch control.
+	//
+	// Note: thrust boost is computed assuming thrust_scale==1 for all motors.
+	// On asymmetric platforms, some motors have thrust_scale<1,
 	// which may result in motor saturation after thrust boost is applied
 	// TODO: revise the saturation/boosting strategy
-
-	if (min_out < 0.0f && max_out < 1.0f && -min_out <= 1.0f - max_out) {
-		float max_thrust_diff = thrust * thrust_increase_factor - thrust;
-
-		if (max_thrust_diff >= -min_out) {
+	if (delta_out_max <= 1.0f) {
+		if (min_out < 0.0f) {
 			boost = -min_out;
 
-		} else {
-			boost = max_thrust_diff;
-			roll_pitch_scale = (thrust + boost) / (thrust - min_out);
-		}
-
-	} else if (max_out > 1.0f && min_out > 0.0f && min_out >= max_out - 1.0f) {
-		float max_thrust_diff = thrust - thrust_decrease_factor * thrust;
-
-		if (max_thrust_diff >= max_out - 1.0f) {
+		} else if (max_out > 1.0f) {
 			boost = -(max_out - 1.0f);
-
-		} else {
-			boost = -max_thrust_diff;
-			roll_pitch_scale = (1 - (thrust + boost)) / (max_out - thrust);
 		}
 
-	} else if (min_out < 0.0f && max_out < 1.0f && -min_out > 1.0f - max_out) {
-		float max_thrust_diff = thrust * thrust_increase_factor - thrust;
-		boost = math::constrain(-min_out - (1.0f - max_out) / 2.0f, 0.0f, max_thrust_diff);
-		roll_pitch_scale = (thrust + boost) / (thrust - min_out);
-
-	} else if (max_out > 1.0f && min_out > 0.0f && min_out < max_out - 1.0f) {
-		float max_thrust_diff = thrust - thrust_decrease_factor * thrust;
-		boost = math::constrain(-(max_out - 1.0f - min_out) / 2.0f, -max_thrust_diff, 0.0f);
-		roll_pitch_scale = (1 - (thrust + boost)) / (max_out - thrust);
+	} else {
+		roll_pitch_scale = 1 / (delta_out_max);
+		boost = 1.0f - ((max_out - thrust) * roll_pitch_scale + thrust);
```

Comment:
I'm not sure what this formula is for...
When min and max are symetric with respect to the middle (thrust) then it's correct but can be simplified to `boost = 0.5f - thrust;`

In an asymetric case which I suppose the math is for the result is incorrect according to my analysis. When I find any mistake I might have made or a better solution I'll suggest it.

Reply:
I found that the calculation here: https://github.com/PX4/Firmware/pull/9062/files#diff-40032af52ac957eae697d7d0a1856ef2R252 only scales roll and pitch (like the name suggests). This makes things unnecessary complicated when the mixer is not symetric.

If it's symetric boosting the thrust to be 0.5 is the right solution because all the authority gets used.

If it's not symetric it would be a lot easier to directly generate the correct output instead of rerunning and rescaling the whole mixer... Similar to what I did here: https://github.com/PX4/Firmware/pull/8485/files#diff-40032af52ac957eae697d7d0a1856ef2R215

Reply:
This formula is for the asymmetric case and is from the constraint 
`(max_out - thrust*thrust_scale)*roll_pitch_scale + (thrust + boost)*thrust_scale = 1.0`. 
Assuming the thrust_scale is 1.0, the constraint is  
`(max_out - thrust)*roll_pitch_scale + thrust + boost = 1.0` 
then 
`boost = 1 - (max_out - thrust)*roll_pitch_scale - thrust`.

It could also be simplified by using the min output:
`(min_out - thrust)*roll_pitch_scale + thrust + boost = 0` 
which gives 
`boost = -(min_out - thrust)*roll_pitch_scale - thrust`

Reply:
@bresch Yeah I found all these variants also with the min:
```
boost1 = 1-thrust-roll_pitch_scale*(max_out - thrust);
boost2 = -thrust -roll_pitch_scale*(min_out - thrust);
```
but my script still gave the wrong output and just now when I wrote with you on slack I realized my script had a mistake in it. Sorry for the churn, it's correct.